File: objectivec/day15_part1_2015.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSArray *lines = [input componentsSeparatedByString:@"\n"];
        
        NSMutableArray *ingredients = [NSMutableArray array];
        
        for (NSString *line in lines) {
            NSArray *parts = [line componentsSeparatedByString:@" "];
            NSInteger capacity = [parts[2] integerValue];
            NSInteger durability = [parts[4] integerValue];
            NSInteger flavor = [parts[6] integerValue];
            NSInteger texture = [parts[8] integerValue];
            NSInteger calories = [parts[10] integerValue];
            
            NSDictionary *ingredient = @{@"capacity": @(capacity), @"durability": @(durability), @"flavor": @(flavor), @"texture": @(texture), @"calories": @(calories)};
            
            [ingredients addObject:ingredient];
        }
        
        NSInteger highestScore = 0;
        
        for (int i = 0; i <= 100; i++) {
            for (int j = 0; j <= 100 - i; j++) {
                for (int k = 0; k <= 100 - i - j; k++) {
                    int l = 100 - i - j - k;
                    
                    NSInteger capacity = MAX(0, i * [ingredients[0][@"capacity"] integerValue] + j * [ingredients[1][@"capacity"] integerValue] + k * [ingredients[2][@"capacity"] integerValue] + l * [ingredients[3][@"capacity"] integerValue]);
                    NSInteger durability = MAX(0, i * [ingredients[0][@"durability"] integerValue] + j * [ingredients[1][@"durability"] integerValue] + k * [ingredients[2][@"durability"] integerValue] + l * [ingredients[3][@"durability"] integerValue]);
                    NSInteger flavor = MAX(0, i * [ingredients[0][@"flavor"] integerValue] + j * [ingredients[1][@"flavor"] integerValue] + k * [ingredients[2][@"flavor"] integerValue] + l * [ingredients[3][@"flavor"] integerValue]);
                    NSInteger texture = MAX(0, i * [ingredients[0][@"texture"] integerValue] + j * [ingredients[1][@"texture"] integerValue] + k * [ingredients[2][@"texture"] integerValue] + l * [ingredients[3][@"texture"] integerValue]);
                    
                    NSInteger score = capacity * durability * flavor * texture;
                    
                    if (score > highestScore) {
                        highestScore = score;
                    }
                }
            }
        }
        
        printf("%ld\n", highestScore);
    }
    return 0;
}

File: haskell/day15_part1_2015.hs
{-# LANGUAGE RecordWildCards #-}

import System.IO (readFile)
import Data.List (maximumBy)
import Data.Ord (comparing)
import Control.Monad (replicateM)

-- | Data structure to represent an ingredient and its properties
data Ingredient = Ingredient
  { capacity    :: Int
  , durability  :: Int
  , flavor      :: Int
  , texture     :: Int
  , calories    :: Int
  } deriving (Show)

-- | Parses a single line of input into an Ingredient
parseIngredient :: String -> Ingredient
parseIngredient line =
  let wordsList = words line
      -- Extract numerical values, ignoring commas and colons
      capacityVal   = read (filter (/= ',') (wordsList !! 2)) :: Int
      durabilityVal = read (filter (/= ',') (wordsList !! 4)) :: Int
      flavorVal     = read (filter (/= ',') (wordsList !! 6)) :: Int
      textureVal    = read (filter (/= ',') (wordsList !! 8)) :: Int
      caloriesVal   = read (wordsList !! 10) :: Int
  in Ingredient capacityVal durabilityVal flavorVal textureVal caloriesVal

-- | Generates all possible allocations of 'total' teaspoons among 'n' ingredients
allocate :: Int -> Int -> [[Int]]
allocate 1 total = [[total]]
allocate n total =
  [ x : xs
  | x <- [0..total]
  , xs <- allocate (n - 1) (total - x)
  ]

-- | Computes the total score for a given allocation of teaspoons
computeScore :: [Int] -> [Ingredient] -> Int
computeScore allocation ingredients =
  let totalCapacity   = max 0 (sum $ zipWith (*) allocation (map capacity ingredients))
      totalDurability = max 0 (sum $ zipWith (*) allocation (map durability ingredients))
      totalFlavor     = max 0 (sum $ zipWith (*) allocation (map flavor ingredients))
      totalTexture    = max 0 (sum $ zipWith (*) allocation (map texture ingredients))
  in totalCapacity * totalDurability * totalFlavor * totalTexture

-- | Main function to read input, process allocations, and determine the maximum score
main :: IO ()
main = do
  -- Read and parse the input file
  content <- readFile "input.txt"
  let ingredientLines = lines content
      ingredients = map parseIngredient ingredientLines
      numIngredients = length ingredients

  -- Generate all possible allocations of 100 teaspoons among the ingredients
  let allocations = allocate numIngredients 100

  -- Compute scores for all allocations
      scores = map (`computeScore` ingredients) allocations

  -- Determine the maximum score
      maxScore = maximum scores

  -- Output the result
  putStrLn $ "The highest possible cookie score is: " ++ show maxScore
